File: 2023/python/day02.py
from help import get_input

MAX = {
    'red': 12,
    'green': 13,
    'blue': 14
}


def main():
    d = get_input('02').strip().split('\n')
    t1 = 0
    t2 = 0
    for l in d:
        g, r = l.split(':', maxsplit=1)
        g = int(g.split(' ')[1])

        # part 1
        g_valid = True
        # part 2
        p2 = {'red': 0, 'green': 0, 'blue': 0}

        for r in r.strip().split('; '):
            parts = r.strip().split(', ')
            parts = [p.split(' ') for p in parts]
            parts = [(p[1], int(p[0])) for p in parts]
            parts = dict(parts)

            for k, v in parts.items():
                # part 1
                if g_valid and MAX[k] < v:
                    g_valid = False
                # part 2
                p2[k] = max(v, p2[k])

        # part 1
        if g_valid:
            t1 += g

        # part 2
        x = 1
        for n in p2.values():
            x *= n
        t2 += x

    print(t1, t2)


if __name__ == '__main__':
    main()

File: 2023/swift/AdventOfCode2023_CmdLine/AdventOfCode2023_CmdLine/day02.swift
//
//  day02.swift
//  AdventOfCode2023_CmdLine
//
//  Created by gary on 04/01/2024.
//

import Foundation

struct Day02 {
    func run() {
        let d = getInput("02")
        let games = parse(d: d)
        let red = 12
        let green = 13
        let blue = 14
        
        var part1 = 0
        var part2 = 0

        for (n, rounds) in games {
            var isValid = true
            var R = 0
            var G = 0
            var B = 0
            for round in rounds {
                let r = round["red", default: 0]
                let g = round["green", default: 0]
                let b = round["blue", default: 0]
                
                if r > red || g > green || b > blue {
                    isValid = false
                }
                R = max(R, r)
                G = max(G, g)
                B = max(B, b)
            }
            if isValid {
                part1 += n
            }
            part2 += R*G*B
        }
        print(part1)
        print(part2)
    }
}

private extension Day02 {
    func parse(d: String) -> [Int: [[String: Int]]] {
        var games: [Int: [[String: Int]]] = [:]
        for line in d.split(separator: "\n") {
            let parts = line.split(separator: ": ")
            let game = Int(parts[0].split(separator: " ")[1])!
            var rounds: [[String: Int]] = []
            for p1 in parts[1].split(separator: "; ") {
                var round: [String: Int] = [:]
                for p2 in p1.split(separator: ", ") {
                    let p3 = p2.split(separator: " ")
                    let colour = String(p3[1])
                    let n = Int(p3[0])
                    round[colour] = n
                }
                rounds.append(round)
            }
            games[game] = rounds
        }
        return games
    }
}
